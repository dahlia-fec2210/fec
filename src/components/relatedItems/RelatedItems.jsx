/* eslint-disable react/no-array-index-key */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable import/extensions */
/* eslint-disable react/prop-types */
import React, { useState, useEffect } from 'react';
import axios from 'axios';
import RelatedProduct from './RelatedProduct.jsx';
import ComparisonModal from './ComparisonModal.jsx';
import OutfitList from '../outfitList/OutfitList.jsx';

const serverRoute = `http://localhost:${process.env.PORT}`;

function RelatedItems({
  currentProduct, setCurrentProduct,
}) {
  const [left, setLeft] = useState(0);
  const [relatedProducts, setRelatedProducts] = useState([]);
  const [openModal, setOpenModal] = useState(false);
  const [modalProduct, setModalProduct] = useState(null);
  const [productData, setProductData] = useState({});
  const [averages, setAverages] = useState({});

  useEffect(() => {
    setLeft(0);
    const cache = JSON.parse(localStorage.getItem('relatedProducts')) || {};
    if (!cache[currentProduct]) {
      axios.get(`${serverRoute}/products/${currentProduct}/related`)
        .then((data) => {
          const uniqueProducts = [];
          const products = data.data;
          products.forEach((product) => {
            if (!uniqueProducts.includes(product) && product !== currentProduct) {
              uniqueProducts.push(product);
            }
          });
          setRelatedProducts(uniqueProducts);
          setLeft(0);
          cache[currentProduct] = uniqueProducts;
          localStorage.setItem('relatedProducts', JSON.stringify(cache));
        });
    } else {
      setRelatedProducts(cache[currentProduct]);
    }
  }, [currentProduct]);

  function shiftRight(event) {
    event.preventDefault();
    setLeft(left + 272);
    axios.post(`${serverRoute}/interactions`, { element: event.target.id, widget: 'Related Products', time: new Date().toTimeString() });
  }

  function shiftLeft(event) {
    event.preventDefault();
    event.preventDefault();
    setLeft(left - 272);
    axios.post(`${serverRoute}/interactions`, { element: event.target.id, widget: 'Related Products', time: new Date().toTimeString() });
  }

  if (relatedProducts.length > 0) {
    return (
      <div>
        <div className="related-container">
          <h4 className="related-heading">Related Products</h4>
          <div className="related-carousel">
            { relatedProducts.length > 0 ? (
              relatedProducts.map((product, index) => (
                <RelatedProduct
                  setOpenModal={setOpenModal}
                  setModalProduct={setModalProduct}
                  relatedProducts={relatedProducts}
                  currentProduct={currentProduct}
                  setCurrentProduct={setCurrentProduct}
                  key={index}
                  left={left}
                  product={product}
                  productData={productData}
                  setProductData={setProductData}
                  averages={averages}
                  setAverages={setAverages}
                />
              ))

            ) : null}
          </div>
          <div className="related-arrows">
            <span className="related-arrow related-arrow-left" onClick={shiftLeft}>{ left === 0 ? null : <i id="related-products-carousel-left-arrow" className="related-icon fa-solid fa-chevron-left fa-2xl" />}</span>
            <span className="related-arrow related-arrow-right" onClick={shiftRight}>{ left >= (relatedProducts.length - 4) * 272 ? null : <i id="related-products-carousel-right-arrow" className="related-icon fa-solid fa-chevron-right fa-2xl" />}</span>
          </div>
        </div>

        {openModal && (
        <ComparisonModal
          modalProduct={modalProduct}
          currentProduct={currentProduct}
          setOpenModal={setOpenModal}
        />
        )}
        <OutfitList
          currentProduct={currentProduct}
          averages={averages}
          setAverages={setAverages}
        />
      </div>
    );
  }
  return null;
}

export default RelatedItems;
