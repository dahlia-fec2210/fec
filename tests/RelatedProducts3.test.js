/* eslint-disable no-undef */
/* eslint-disable import/extensions */
/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { setupServer } from 'msw/node';
import { rest } from 'msw';
import renderer from 'react-test-renderer';

import RelatedItems from '../src/components/relatedItems/RelatedItems.jsx';

const mockResponse = [37316, 37318, 37319, 37311, 37313];

const handlers = [
  rest.get('http://localhost:3001/products/:productId/related', (req, res, ctx) => {
    if (req.params.productId === '37315') {
      return res(ctx.json(mockResponse));
    }
    return res(ctx.json({ boo: 'error' }));
  }),
];

const server = setupServer(...handlers);

beforeAll(() => server.listen());
beforeAll(() => server.resetHandlers());
afterAll(() => server.close());

test('renders related products and outfit list with 37315', async () => {
  const tree = renderer
    .create(<RelatedItems currentProduct={37315} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
