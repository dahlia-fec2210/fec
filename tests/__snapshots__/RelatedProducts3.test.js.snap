// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders related products and outfit list with 37315 1`] = `
<div>
  <div
    className="related-container"
  >
    <h4
      className="related-heading"
      data-testid="related-heading"
    >
      Related Products
    </h4>
    <div
      className="related-carousel"
    />
    <div
      className="related-arrows"
    >
      <span
        className="related-arrow related-arrow-left"
        data-testid="related-arrow-left"
        onClick={[Function]}
      />
      <span
        className="related-arrow related-arrow-right"
        data-testid="related-arrow-right"
        onClick={[Function]}
      />
    </div>
  </div>
  <div>
    <div
      className="related-container"
    >
      <h4
        className="related-heading"
        data-testid="outfit-heading"
      >
        Your Outfit
      </h4>
      <div
        className="outfit-row"
      >
        <div
          className="related-add-card"
        >
          <div
            className="related-add-outfit-heading"
            data-testid="add-card"
          >
            Add to Outfit
          </div>
          <div
            className="related-stack"
            data-testid="add-to-outfit"
            onClick={[Function]}
          >
            <div
              className="fa-stack"
              style={
                {
                  "verticalAlign": "top",
                }
              }
            >
              <i
                className="related-circle fa-solid fa-regular fa-circle fa-stack-2x"
              />
              <i
                className="related-star fa-solid fa-plus fa-stack-1x"
              />
            </div>
          </div>
        </div>
        <div
          className="outfit-carousel"
          data-testid="outfit-carousel"
        />
      </div>
      <div
        className="related-arrows"
      >
        <div
          className="related-arrow related-arrow-left"
          data-testid="outfit-arrow-left"
          onClick={[Function]}
        />
        <div
          className="related-arrow related-arrow-right"
          data-testid="outfit-arrow-right"
          onClick={[Function]}
        />
      </div>
    </div>
  </div>
</div>
`;
